node/authorship/src/authorship.rs

```diff
@@ -520,25 +548,87 @@ where
 
         self.transaction_pool.remove_invalid(&unqueue_invalid);
 
-        // Pushing a custom extrinsic that must run at the end of a block
-        let custom_extrinsic = block_builder.create_terminal_extrinsic(self.max_gas)?;
+        // Attempt to apply pseudo-inherent on top of the current overlay in a separate thread.
+        // In case the timeout is hit, we will proceed without it.
+        let client = self.client.clone();
+        let backend = self.backend.clone();
+        let pseudo_inherent = block_builder.create_terminal_extrinsic(self.max_gas)?;
+        let (extrinsics, api, version, _, _, estimated_header_size) =
+            block_builder.clone().deconstruct();
+        // We need the overlay changes and transaction storage cache to send to a new thread.
+        // The cloned `RuntimaApi` object can't be sent to a new thread directly so we have to
+        // break it down into parts (that are `Send`) and then reconstruct it in the new thread.
+        // If changes applied successfully, the updated extrinsics and api parts will be sent back
+        // to update the original block builder and finalize the block.
+        let (_, api_params) = api.deref().clone().deconstruct();
+
+        let update_block = async move {
+            let (tx, rx) = oneshot::channel();
+            let spawn_handle = self.spawn_handle.clone();
+
+            spawn_handle.spawn_blocking(
+                "block-builder-push",
+                None,
+                Box::pin(async move {
+                    debug!(target: "gear::authorship", "⚙️  Pushing Gear::run extrinsic into the block...");
+                    let mut block_builder = BlockBuilder::<'_, Block, C, B>::from_parts(
+                        extrinsics,
+                        ApiRef::from(C::Api::restore(client.as_ref(), api_params)),
+                        version,
+                        parent_hash,
+                        backend.as_ref(),
+                        estimated_header_size);
+                    let outcome = block_builder.push(pseudo_inherent).map(|_| {
+                        let (extrinsics, api, _, _, _, _) =
+                            block_builder.deconstruct();
+                        let (_, api_params) = api.deref().clone().deconstruct();
+                        (extrinsics, api_params)
+                    });
+                    if tx.send(outcome).is_err() {
+                        debug!(
```

Comment:
May be error or warn log here?

Reply:
I'd prefer `log::warn!` because we have already logged the error by this moment in time on the receiving end (when the timeout was fired). Here it only indicates that the thread has finished the computations but the result is no longer needed (channel closed).